src/time.cpp

```diff
@@ -17,118 +17,167 @@
+    {Boundaries, "copying boundaries"},
+    {MpiAllTime, "all-all communication"},
+    {MpiOneTime, "1-1 communication"},
+    {FieldOutput, "outputting fields"},
+    {FourierTransforming, "Fourier transforming"},
+    {MPBTime, "MPB mode solver"},
+    {GetFarfieldsTime, "far-field transform"},
+    {FieldUpdateB, "updating B field"},
+    {FieldUpdateH, "updating H field"},
+    {FieldUpdateD, "updating D field"},
+    {FieldUpdateE, "updating E field"},
+    {BoundarySteppingB, "boundary stepping B"},
+    {BoundarySteppingWH, "boundary stepping WH"},
+    {BoundarySteppingPH, "boundary stepping PH"},
+    {BoundarySteppingH, "boundary stepping H"},
+    {BoundarySteppingD, "boundary stepping D"},
+    {BoundarySteppingWE, "boundary stepping WE"},
+    {BoundarySteppingPE, "boundary stepping PE"},
+    {BoundarySteppingE, "boundary stepping E"},
+    {Other, "everything else"},
+};
+
+std::vector<double> timing_data_vector(const std::unordered_map<time_sink, double> &timers) {
+  std::vector<double> ret;
+  for (const auto &desc_ts : kDescriptionByTimeSink) {
+    auto it = timers.find(desc_ts.first);
+    ret.push_back((it != timers.end()) ? it->second : 0.);
+  }
+  return ret;
+}
+
+std::vector<double> timing_data_vector_from_all(const std::unordered_map<time_sink, double> &timers) {
+  std::vector<double> time_spent_vector = timing_data_vector(timers);
+  const int n = count_processors();
+  std::vector<double> alltimes_tmp(n * time_spent_vector.size());
+  for (size_t i = 0; i < time_spent_vector.size(); ++i) {
+    alltimes_tmp[i * n + my_rank()] = time_spent_vector[i];
+  }
+
+  std::vector<double> alltimes(alltimes_tmp.size());
+  sum_to_all(alltimes_tmp.data(), alltimes.data(), alltimes_tmp.size());
+  return alltimes;
+}
+
+void pt(double mean, double stddev, const char *label) {
+  if (mean != 0) {
+    if (stddev != 0)
+      master_printf("    %21s: %g s +/- %g s\n", label, mean, stddev);
+    else
+      master_printf("    %21s: %g s\n", label, mean);
+  }
+}
+
+} // namespace
+
+timing_scope::timing_scope(std::unordered_map<time_sink, double> *timers_, time_sink sink_)
+    : timers(timers_), sink(sink_), active(true), t_start(wall_time()) {}
+
+timing_scope::~timing_scope() { exit(); }
+
+void timing_scope::exit() {
+  if (!active) return;
+  (*timers)[sink] += (wall_time() - t_start);
+  active = false;
+}
+
+timing_scope fields::with_timing_scope(time_sink sink) { return timing_scope(&times_spent, sink); }
+
 void fields::finished_working() {
-  double now = wall_time();
-  if (last_wall_time >= 0) times_spent[working_on] += now - last_wall_time;
-  last_wall_time = now;
-  working_on = was_working_on[0];
-  for (int i = 0; i < MEEP_TIMING_STACK_SZ - 1; ++i)
-    was_working_on[i] = was_working_on[i + 1];
-  was_working_on[MEEP_TIMING_STACK_SZ - 1] = Other;
+  if (!was_working_on.empty()) { was_working_on.pop_back(); }
+  working_on =
+      with_timing_scope(!was_working_on.empty() ? was_working_on.back() : Other);
```

Comment:
Don't you need to call `pop_back` after you call `was_working_on.back()`, rather than before?

Reply:
Oh, I see that `was_working_on` actually includes the current `sink` too, which is why you need to pop first.